src/config.c

```diff
@@ -5,13 +5,145 @@
 
+/*
+ *  construct_parent_directories()
+ *
+ *  Returns:
+ *  0 Success
+ * -1 Failure
+ */
+int create_parent_directories(char *path)
+{
+	/*
+	 * We use the `p` pointer to find each '/', then we temporary replace the '/'
+	 * character with '\0', so we can actually create the directory.
+	 * After that the '\0' gets replaced with '/' so we can continue iterating
+	 * through the path. The final directory won't be created this way, therefore
+	 * we have another call to `mkdir` after our `for` loop to make sure all the
+	 * directories are created.
+	 */
+	char *p;
+	for (p = path + 1; *p; p++) {
+		if (*p == '/') {
+			*p = '\0';
+			if (mkdir(path, 0777) != 0) {
+				if (errno != EEXIST) {
+					if (errno == EACCES) {
+						pr_error(
+							"Error creating directory %s, permission denied\n",
+							path);
+						return -1;
+					}
+				}
+			}
+			*p = '/';
+		}
+	}
+
+	if (mkdir(path, 0777) != 0) {
+		if (errno != EEXIST) {
+			if (errno == EACCES) {
+				pr_error(
+					"Error creating directory %s, permission denied\n",
+					path);
+				return -1;
+			}
+		}
+	}
+
+	// iterate to the end of the path so we can make sure to append '/' if needed
+	while (*p) {
+	}
+	if (*(p - 1) != '/') {
+		strncat(path, "/", strlen(path) + 1);
+	}
+
+	return 0;
+}
+
+/*
+ *  construct_log_filepath()
+ *
+ *  Returns:
+ *  0 Success
+ * -3 Failure
+ */
+int construct_log_filepath(void)
+{
+	struct stat st = { 0 };
+	int ret = 0;
+
+	// If the user didn't specify a path, we create the `cnc` directory under `/var/log/`
+	if (ini_config->general_config->log_filepath == NULL ||
+	    strcmp(ini_config->general_config->log_filepath, "") == 0) {
+		/*
+		 * If `stat` returns `-1`, it means that we couldn't find the directory,
+		 * therfore we create it.
+		 */
+		ret = stat(LOG_FILEPATH, &st);
+		if (ret == -1) {
+			ret = mkdir(LOG_FILEPATH, 0777);
+			if (ret != 0) {
+				/*
+				 * If we can't create the directory due to access rights,
+				 * we default to `$HOME/`
+				 */
+				if (errno == EACCES) {
+					pr_warn("No permissions to create directory on `/var/log/`. "
+						"Current directory will be used.\n");
+
+					char current_directory[256];
+					if (getcwd(current_directory,
+						   sizeof(current_directory)) !=
+					    NULL) {
+						strncat(current_directory, "/",
+							sizeof(current_directory));
+						log_filepath = strdup(
+							current_directory);
+					}
+				}
+			} else {
+				log_filepath = strdup(LOG_FILEPATH);
+			}
+		} else {
+			log_filepath = strdup(LOG_FILEPATH);
+		}
+	} else {
+		// else we iterate through the path and create each parent directory needed
+		char path[256];
+		snprintf(path, sizeof(path), "%s",
+			 ini_config->general_config->log_filepath);
+		if (*path != '/') {
+			pr_error(
+				"Please provide an absolute path for log filepath\n");
+			return -3;
+		}
+
+		if (create_parent_directories(path) != 0) {
+			return -3;
+		}
+
+		log_filepath = strdup(path);
+	}
+
+	return 0;
+}
```

Comment:
I suggest this, with checking for successful `strdup` becoming a separated function to use all across our code base.

```suggestion
/*
 *  construct_log_filepath()
 *
 *  Returns:
 *  0 Success
 * -3 Failure
 */
int construct_log_filepath(void)
{
	struct stat st = { 0 };
	int ret = 0;

	// If the user didn't specify a path, we create the `cnc` directory under `/var/log/`
	if (ini_config->general_config->log_filepath == NULL ||
	    strcmp(ini_config->general_config->log_filepath, "") == 0) {
		ret = stat(LOG_FILEPATH, &st);
		if (ret == -1) {
			ret = mkdir(LOG_FILEPATH, 0777);
			if (ret != 0) {
				if (errno == EACCES) {
					pr_warn("No permissions to create directory on `/var/log/`. "
						"Current directory will be used.\n");

					char current_directory[256];
					if (getcwd(current_directory,
						   sizeof(current_directory)) !=
					    NULL) {
						size_t len = strlen(
							current_directory);
						if (len <
						    sizeof(current_directory) -
							    2) {
							strncat(current_directory,
								"/",
								sizeof(current_directory) -
									len -
									1);

							char *temp_log_filepath = strdup(
								current_directory);
							if (temp_log_filepath !=
							    NULL) {
								log_filepath =
									temp_log_filepath;
							} else {
								pr_error(
									"Failed to allocate memory for log filepath.\n");
								return -3;
							}
						}
					}
				}
			} else {
				char *temp_log_filepath = strdup(LOG_FILEPATH);
				if (temp_log_filepath != NULL) {
					log_filepath = temp_log_filepath;
				} else {
					pr_error(
						"Failed to allocate memory for log filepath.\n");
					return -3;
				}
			}
		} else {
			char *temp_log_filepath = strdup(LOG_FILEPATH);
			if (temp_log_filepath != NULL) {
				log_filepath = temp_log_filepath;
			} else {
				pr_error(
					"Failed to allocate memory for log filepath.\n");
				return -3;
			}
		}
	} else {
		// else we iterate through the path and create each parent directory needed
		char path[256];
		snprintf(path, sizeof(path), "%s",
			 ini_config->general_config->log_filepath);
		if (*path != '/') {
			pr_error(
				"Please provide an absolute path for log filepath\n");
			return -3;
		}

		if (create_parent_directories(path) != 0) {
			return -3;
		}

		char *temp_log_filepath = strdup(path);
		if (temp_log_filepath != NULL) {
			log_filepath = temp_log_filepath;
		} else {
			pr_error(
				"Failed to allocate memory for log filepath.\n");
			return -3;
		}
	}

	return 0;
}
```

Reply:
Which part should be a seperate function?

Reply:
The checking for strdup with a temp variable.